Torch Dataset

kalebphipps opened the pull request:
# Description

This PR includes a ``torch.Datset`` for Calibration Items from the ``PAINT`` database. This includes:
- A dataset that can be initialized with a ``root_dir`` containing calibration items.
- The possibility to load different calibration items, i.e. raw images, cropped images, flux images, flux_centered images, or calibration properties.
- The possibility to initialize a dataset from a benchmark dataset ``csv`` file containing information on the splits.
- The possibility to initialize a dataset using all calibration items from one or more heliostats.

To enable minimal tests, some data files (images and JSONs) have been uploaded into Git.

This PR therefore, closes Issue #10.

In addition, unused variables were renamed to clear up any confusion from Issue #59. 

**NOTE: Due to download functions in the dataset that are not tested, this PR will decrease code coverage!**


## Type of change

<!-- Please delete options that are not relevant. -->

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules

Markus-Goetz commented on tests/data/test_data/dataset/validation/106293-calibration-properties.json:
Should this file be pretty printed, i.e. correct indentations and line breaks?

Markus-Goetz commented on tests/data/test_data/dataset/validation/213508-calibration-properties.json:
Should this file be pretty printed, i.e. correct indentations and line breaks?

Markus-Goetz commented on tests/data/test_data/dataset/validation/99927-calibration-properties.json:
Should this file be pretty printed, i.e. correct indentations and line breaks?

Markus-Goetz commented on tests/data/test_data/test_benchmark.csv:
The column headers have inconsistent capitalization. I assume that it does not majorly matter for the code later. However, I believe it may be worthwhile to make these consistent, e.g. change id to ID or Id

kalebphipps replied:
I have changed id to ID and fixed the instances where this caused errors!

Markus-Goetz commented on paint/util/paint_mappings.py:
```diff
@@ -452,3 +452,11 @@
+CALIBRATION_CROPPED_IMAGE_IDENTIFIER = "_cropped.png"
+CALIBRATION_FLUX_IMAGE_IDENTIFIER = "_flux.png"
+CALIBRATION_FLUX_CENTERED_IMAGE_IDENTIFIER = "_flux_centered.png"
+CALIBRATION_PROPERTIES_IDENTIFIER = "-calibration-properties.json"
```
This is inconsistent now. Calibration has dashes, while all the other have underscores. Should it remain this way or do we want to unify the names to dashes only (to avoid OS-specific problems)?

Markus-Goetz commented on paint/preprocessing/focal_spot_extractor.py:
```diff
@@ -30,7 +30,9 @@ def __init__(self) -> None:
         self.aim_point_image: tuple[float, float] = (0.0, 0.0)  # Default to (0.0, 0.0)
-        self.aim_point: torch.Tensor = torch.empty(3)  # Default to an empty tensor of required shape
+        self.aim_point: torch.Tensor = torch.empty(
+            3
```
Very odd line break due to the long comment. Should this go to the next line?

Markus-Goetz commented on scripts/example_benchmark_dataset_full_workflow.py:
```diff
@@ -0,0 +1,118 @@
+        help="Root directory to save outputs.",
+        default=f"{PAINT_ROOT}/benchmarks",
+    )
+    parser.add_argument(
```
This probably should include the choices keyword https://docs.python.org/3/library/argparse.html#choices

kalebphipps replied:
I have included choice arumgnets

Markus-Goetz commented on scripts/example_benchmark_dataset_full_workflow.py:
```diff
@@ -0,0 +1,118 @@
+    parser.add_argument(
+        "--train_size",
+        type=int,
+        help="The size of the training split.",
```
in percentage?

kalebphipps replied:
In number of samples - I have adjusted this in the description!

Markus-Goetz commented on scripts/example_benchmark_dataset_full_workflow.py:
```diff
@@ -0,0 +1,118 @@
+    parser.add_argument(
+        "--remove_unused_data",
+        type=bool,
+        help="Whether to remove unused data or not.",
```
What is unused data? Can we be more specific here?

kalebphipps replied:
I have included a more informative description!

Markus-Goetz commented on paint/data/dataset.py:
```diff
@@ -0,0 +1,437 @@
+from torch.utils.data import Dataset
+from torchvision import transforms
+from tqdm import tqdm
+from typing_extensions import Self
```
Should this be selected based on the Python version? If so, everything before 3.11 should use typing_extensions and everything afterwards should use typing

kalebphipps replied:
Done :)

Markus-Goetz commented on paint/data/dataset.py:
```diff
@@ -0,0 +1,437 @@
+        self.item_ids = item_ids
+        self.to_tensor = transforms.ToTensor()
+
+    def __str__(self) -> str:
```
I guess this can be easily tested (not sure if necessary)

kalebphipps replied:
I have included a test for this

Markus-Goetz commented on paint/data/dataset.py:
```diff
@@ -0,0 +1,437 @@
+        if key not in accepted_items:
+            raise ValueError(
+                f"The calibration type {key} is not accepted. Please select one of "
+                f"{mappings.CALIBRATION_RAW_IMAGE_KEY}, {mappings.CALIBRATION_CROPPED_IMAGE_KEY}, "
```
Do you want to just use *accepted_keys here?

kalebphipps replied:
Good idea, I have changed this

Markus-Goetz commented on paint/data/dataset.py:
```diff
@@ -0,0 +1,437 @@
+        str
+            Mapped key to the file identifier.
+        """
+        mapper = {
```
This will be reconstructed over and over. Should this be a constant variable of the class instead?

kalebphipps replied:
Good idea, II have changed this

Markus-Goetz commented on paint/data/dataset.py:
```diff
@@ -0,0 +1,437 @@
+
+        # Initialize a train dataset.
+        train_dataset = cls(
+            root_dir=Path(root_dir) / mappings.TRAIN_INDEX,
```
root_dir is initialized multiple times as a pathlib.Path here. Possibly just do it once?

kalebphipps replied:
Fixed

Markus-Goetz commented on paint/data/dataset.py:
```diff
@@ -0,0 +1,437 @@
+
+        # Initialize a test dataset.
+        test_dataset = cls(
+            root_dir=Path(root_dir) / mappings.TEST_INDEX,
```
root_dir is initialized multiple times as a pathlib.Path here. Possibly just do it once?

Markus-Goetz commented on paint/data/dataset.py:
```diff
@@ -0,0 +1,437 @@
+
+        # Initialize a validation dataset.
+        validation_dataset = cls(
+            root_dir=Path(root_dir) / mappings.VALIDATION_INDEX,
```
root_dir is initialized multiple times as a pathlib.Path here. Possibly just do it once?

Markus-Goetz commented on paint/data/dataset.py:
```diff
@@ -0,0 +1,437 @@
+        # Create STAC client.
+        client = StacClient(output_dir=root_dir)
+        ids_dict = {}
+        for split_name, split_data in splits.groupby(mappings.SPLIT_KEY):
```
Purposefully exclude these lines from testing/coverage?

Markus-Goetz commented on paint/data/dataset.py:
```diff
@@ -0,0 +1,437 @@
+            )
+        if download:
+            log.info("Beginning to download the data...")
+            cls._download_heliostat_data(
```
Exclude these lines from testing/coverage?

Markus-Goetz commented on paint/data/dataset.py:
```diff
@@ -0,0 +1,437 @@
+            mappings.CALIBRATION_FLUX_CENTERED_IMAGE_IDENTIFIER,
+        ]:
+            item = cv2.imread(str(file_name))
+            item = self.to_tensor(item)
```
Do you want to use the torchvision.ToTensor() transform here? This might avoid double allocation of memory

kalebphipps replied:
Done

kalebphipps commented on scripts/example_benchmark_dataset_full_workflow.py:
```diff
@@ -0,0 +1,118 @@
+
+    # Determine whether to download the data or not:
+    # The first time this script is executed locally, the data must be downloaded, afterward no longer.
+    if dataset_output_dir.exists():
```
I think it needs to be ``dataset_download = not dataset_output_dir.exists()`` since we want to download when the data is not there - but I have changed this :)

kalebphipps commented on paint/data/dataset.py:
```diff
@@ -0,0 +1,437 @@
+import paint.util.paint_mappings as mappings
+from paint.data import StacClient
+
+log = logging.getLogger(__name__)  # Logger for the STAC client
```
The way the logger is set up is that it actually uses the name to differentiate between what class is logging something. So at the moment we get output like this:

```
[2024-12-09 17:11:47,112][paint.data.dataset_splits][INFO] - Azimuth split complete!
[2024-12-09 17:11:47,192][paint.data.dataset][INFO] - Begining the process of generating benchmark datasets. The file used to generate the benchmarks is:
```

I would therefore prefer to keep it as is for now :)

kalebphipps commented:
I have addressed all comments and either made comments explaining why I don't plan on making the change or made the change directly and resolved the conversation.

The coverage issue has been resolved by excluding certain lines (i.e. download functions) from the coverage and including an additional test to verify the string representation. I have also completely reworked the ``README.md`` to highlight the new repository structure and example scripts.
